### ISSUE TITLE
Getting errno "bad file descriptor" writing to /dev/osxfuse0 on occassion

### ISSUE THREAD

**Issue by edmc-ss** [osxfuse/osxfuse#671]
Caveat: Not using user-mode (C) library as my application is in Golang (see https://github.com/swiftstack/fission) - this is strictly attempting to talk via FUSE device file to OSXFuse's .kext

In the aforementioned Golang package (fission), there is an example (fission-ramfs) that you can easily run. Just compile with Golang v1.13+ on your Mac. The program takes a single argument... an empty directory... where the file system (backed only by RAM) will be mounted.

Simply executing `dd if=/dev/zero of=mount_point/foo bs=64k count=8` will reliably reproduce the problem. The "bad file descriptor" error happens on the OpFlush upcall's response... after all 8 of the 64KiB OpWrite's have executed... and dozens of other Op's prior.

Confusing why this would be happening... Any suggestions on how to debug the issue?

**Comment by edmc-ss** [comment]
Not making much progress on this... source code would be nice :-).

I notice that the two Bazil FUSE examples are read-only so no help there...

Looking at github.com/hanwen/go-fuse & github.com/jacobsa/fuse for hopefully better examples that might either exhibit the same problem or show a different usage of OSXFuse's .kext that doesn't fail in this way...

Meanwhile, I've verified the failure happens against various recent OSXFuse versions (3.10.x) but as my development Mac is on Catalina, I believe I am stuck at using only these latest (closed source) versions, no?

Here is my mount sequence:
```
	iosizeMountOption = fmt.Sprintf("iosize=%d", volume.initOutMaxWrite)

	mountOptions = iosizeMountOption

		volume.devFuseFD, err = syscall.Open(devOsxFusePath, syscall.O_RDWR|syscall.O_CLOEXEC, 0)
		osxFuseMountCmd = &exec.Cmd{
			Path: osxFuseMountPath,
			Args: []string{
				osxFuseMountPath,
				"-o", mountOptions,
				"3",
				volume.mountpointDirPath,
			},
			Env:          append(os.Environ(), osxFuseMountCallByEnv, osxFuseDaemonPathEnv+os.Args[0]),
			Dir:          "",
			Stdin:        nil,
			Stdout:       nil, // This will be redirected to osxFuseMountCmdCombinedOutput below
			Stderr:       nil, // This will be redirected to osxFuseMountCmdCombinedOutput below
			ExtraFiles:   []*os.File{os.NewFile(uintptr(volume.devFuseFD), devOsxFusePath)},
			SysProcAttr:  nil,
			Process:      nil,
			ProcessState: nil,
		}

		osxFuseMountCmdCombinedOutput, err = osxFuseMountCmd.CombinedOutput()
```
In the above, I'm launching the devFuseFDReader() that ultimately dispatches the Init() handler. devFuseFD is the open'd /dev/osxfuse0 file (I'm the only user at this point). That CombinedOutput() method on the cmd basically says "run it and collect stdout/stderr in a big string"...

Later, when I issue 8 128KiB writes, the 8th one gets the device closed error I'm asking about. Any number of requests have been successfully responded to prior (including the prior 7 writes of course).

I suppose the other pertinent info would be how I respond to Init():
```
const 	initOutFlagsMask            = fission.InitFlagsAsyncRead | fission.InitFlagsBigWrites | fission.InitFlagsDontMask | fission.InitFlagsAutoInvalData | fission.InitFlagsDoReadDirPlus
...
func (dummy *globalsStruct) DoInit(inHeader *fission.InHeader, initIn *fission.InitIn) (initOut *fission.InitOut, errno syscall.Errno) {
	initOut = &fission.InitOut{
		Major:                initIn.Major,
		Minor:                initIn.Minor,
		MaxReadAhead:         initIn.MaxReadAhead,
		Flags:                (initIn.Flags & initOutFlagsMask) | fission.InitFlagsBigWrites,
		MaxBackground:        initOutMaxBackgound,
		CongestionThreshhold: initOutCongestionThreshhold,
		MaxWrite:             initOutMaxWrite,
	}

	errno = 0
	return
}
```

**Comment by edmc-ss** [comment]
To duplicate:
1) Install Golang
2) Locate or create a dir for Golang work (e.g. ~/go)
3) export GOPATH=~/go (or whereever you put it)
4) export PATH=$PATH:$GOPATH/bin
5) mkdir -p $GOPATH/src/github.com/swiftstack
6) cd $GOPATH/src/github.com/swiftstack
7) git clone https://github.com/swiftstack/fission.git
8) cd fission/examples/fission-ramfs
9) go install .
10) mkdir mount_point
11) fission-ramfs mount_point
12) [In another terminal] dd if=/dev/zero of=mount_point/foo bs=128k count=8

**Comment by edmc-ss** [comment]
Thanks for taking a look at this Benjamin...
In the meantime, I've found an even easier way to trigger the anomaly:
    dd if=/dev/zero of=mount_point/foo bs=128k count=1 seek=6
Yup - all I'm doing is asking dd to seek 6 "blocks" (i.e. 6 * 128KiB) into the new file before doing a single 128KiB write. The SetAttr works just fine... but the immediately following Write (with the offset == the size the file was set to via the SetAttr call) is what gets the "bad file descriptor" error. Very strange...

**Comment by edmc-ss** [comment]
Very discouraging lack of progress on this significant failure.

Background: I developed package fission precisely to multi-thread the FUSE upcall handling (as was not adequately done by at least Bazil FUSE but also, I claim, the other two Golang libraries for doing the same3).

Anyway, I just disabled the multi-threading behavior and the "bad file descriptor" errors became much harder to replicate. Alas, they still occur.

Does anyone have at least the pre-closed-source of the .kext of OSXFuse that I can take a look at. I fear only Benjamin can debug this w/out others having .kext source code access :-(.

**Comment by edmc-ss** [comment]
The aforementioned `dd` command, if seeking to 8 (128KiB) blocks instead of 6, and converting my package fission to be single threaded results in the "bad file descriptor" error immediately... and it happens on the reader! Previously, the (first) "bad file descriptor" error would be hit by the writer... either on the setattr (we are seeking/extending the file size after all), the write itself, of the subsequent "flush" call (when the file is closed by `dd`).

Is there, perhaps, some problem with OSXFuse's .kext having a read outstanding on the /dev/osxfuse* file at the time of the write (for a previous upcall)?

**Comment by edmc-ss** [comment]
Even simpler repro...
TERM 1: `fission-ramfs mount_point`
TERM 2: `for i in `seq 5000`;do stat mount_point >/dev/null 2>&1;done`
After about 4800 stats (YMMV), the "bad file descriptor" error will regularly hit

**Comment by edmc-ss** [comment]
I've added a "hellofs" example to my Golang package "fission" to keep what I'm emulating as simple as possible - and to match a typical example FUSE file system popular in Linux's FUSE, OSXFuse, and the 3 other Golang wrappers around the user mode side of FUSE.

Still the same result is very possible... but I was able to notice one very interesting thing. I added code to dump out what is received and returned for a Access() upcall from FUSE... and I see the following:

UNDO: DoAccess(inHeader: &fission.InHeader{Len:0x30, OpCode:0x22, Unique:0x3, NodeID:0x1, UID:0x1f6, GID:0x14, PID:0xf496, Padding:0x0}, accessIn: &fission.AccessIn{Mask:0x0, Padding:0x0})... ...returning errno: errno 0
UNDO: DoAccess(inHeader: &fission.InHeader{Len:0x30, OpCode:0x22, Unique:0x2, NodeID:0x1, UID:0x1f6, GID:0x14, PID:0xf497, Padding:0x0}, accessIn: &fission.AccessIn{Mask:0x0, Padding:0x0})... ...returning errno: errno 0

In other words, I'm getting upcalls for doing Access() on NodeID 1 (the root of the volume) with a Mask of Zero: so-called "F_OK" which is simply asking "does this Inode exist. As you can see, I'm returning 0 saying "well, yes, it does".

That's not the strange part. The strange part is that these two Access() upcalls, with Unique values of 2 and 3 (not sure why there are two of them) are repeated about every 10 seconds!!!

This is obviously not good... but I think the kernel may just be re-issuing the request - REPEATING THE USE OF UNIQUE - for some reason. Understandably, the OSXFuse .kext may be upset and ultimately just close my open handle on the /dev/osxfuse* file.

Possible?

**Comment by edmc-ss** [comment]
Apparently, there is some program "polling" the mount point:
  /System/Library/PrivateFrameworks/SyncedDefaults.framework/Support/syncdefaultsd

**Comment by edmc-ss** [comment]
Ok... so I have verified that the InHeader coming up from OSXFuse .kext is definitely *NOT* giving me UNIQUE IDs. Kinda makes it hard to respond to them e.g. out of order.

Is this true? Is OSXFuse basically not setting the UNIQUE field in the InHeader and hence ignoring the UNIQUE field in the returned OutHeader?

Here is my Golang declaration of the InHeader:

type InHeader struct {
	Len     uint32 // includes InHeaderSize and any payload
	OpCode  uint32
	Unique  uint64
	NodeID  uint64
	UID     uint32
	GID     uint32
	PID     uint32
	Padding uint32
}

Here is the corresponding declaration of the OutHeader:

type OutHeader struct {
	Len    uint32
	Error  int32
	Unique uint64
}

Of course I have "unsafe" code marshaling the above. Here is the unmarshaling of InHeader:

	inHeader = &InHeader{
		Len:     *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[0])),
		OpCode:  *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[4])),
		Unique:  *(*uint64)(unsafe.Pointer(&devFuseFDReadBuf[8])),
		NodeID:  *(*uint64)(unsafe.Pointer(&devFuseFDReadBuf[16])),
		UID:     *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[24])),
		GID:     *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[28])),
		PID:     *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[32])),
		Padding: *(*uint32)(unsafe.Pointer(&devFuseFDReadBuf[36])),
	}

And here is the marshaling of OutHeader:

	*(*uint32)(unsafe.Pointer(&outHeader[0])) = uint32(iovecSpan) // Updated later
	*(*int32)(unsafe.Pointer(&outHeader[4])) = -int32(errno)
	*(*uint64)(unsafe.Pointer(&outHeader[8])) = inHeader.Unique

The above marshalling code works perfectly under Linux who's .kext equivalent faithfully will give me multiple commands at a time (baring some unfortunate serialization it incorrectly assumes my file system cannot handle such as two writes to the same file at the same time). As I mentioned, OSXFuse does not seem to honor the meaning of the "Unique" field and, despite that, will gladly issue multiple upcalls to me (though it would appear I may be getting only one at a time... unsure).

This sure looks like an important issue. I have absolutely no notion that it is related to why it would appear the OSXFuse .kext is closing the /dev/osxfuse* out from under me.

Would really appreciate some response/feedback from Benjamin here...

**Comment by edmc-ss** [comment]
As I noted in Issue #674, it could be that the reuse of Unique values is actually "ok" so long as they aren't reused until after the user mode agent has responded to their prior use. This does appear to be my case at least...

So I keep looking... guessing at what Benjamin could be thinking here. He has opted not to address this issue at this time since it "only affects my code" alas.

We'll be sending him a SOW with proposed pay rate in the hopes of resolving this issue.

**Comment by edmc-ss** [comment]
I thought I'd try yet again to narrow down the issue... and I think I have some interesting evidence of what is going on.

First of all, I implemented an example "fission-hellofs" that, as other equivalents have done, is a read only file system with a single file ("Hello") in it that can be read.

Anyway, since the whole point of package fission is to be able to multithread, I thought the problem OSXFuse is seeing simply must be related to that... so I have been thinking about what exactly my code must be doing that is unique.

Sure enough, if user mode code is only single threaded, a single read buffer may be used when reading from the /dev/fuse device. So I started playing with that.

First tidbit - by varying the MaxWrite parameter (I default to 128KiB), I can make the failure happen more quickly (larger buffers increase failure rate).

Next, I simply cap'd the number of unique buffers I'm using to read (Go chan's make this really easy)... and, sure enough, as I increase the number of unique buffers, the error rate goes down! It's not the first time I've reused the buffer that causes the failure... more like the 2000th time.

This is looking suspiciously like a buffer reuse race condition **somewhere**...

**Comment by edmc-ss** [comment]
I have finally resolved the issue... and I dare say it was entirely my fault as Benjamin suspected. Indeed, I doubt there would have been any evidence of the issue visible to him debugging the .kext.

To summarize the error, I was constructing an *os.File in Golang to pass to the way you launch a subprocess to execute the equivalent of fusermount... the lifetime of this *os.File would eventually fall out of context and Golang's GC would happily "close" the *os.File and, hence, the underlying Fd I obtained from calling syscall.Open().

The resolution was to instead call os.OpenFile()... and keep that *os.File open and just use the Fd() method to fetch the underlying Fd whenever needed (i.e. when later calling SYSCALL.READ and SYSCALL.WRITEV.

Anyway, you can see the fix in action with the now very performant Golang package fission v1.2.0 ([](https://github.com/swiftstack/fission)) that happily talks to OSXFuse's .kext for full-on multi-threaded FUSE serving.

**Comment by bfleischer** [comment]
Thanks for the update. This might prove very helpful to others.
